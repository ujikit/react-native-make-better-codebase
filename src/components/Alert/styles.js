//package import
import { StyleSheet, StatusBar, Platform } from 'react-native';

//local import
import { COLORS } from '../../configs';
import { FONT_HEADLINE_H6, FONT_BODY5 } from '../../configs/fonts';

export default StyleSheet.create({
  wrapAlert: {
    width: '100%',
    padding: 16,
    position: 'absolute',
    top: Platform.OS === 'android' ? StatusBar.currentHeight : 50,
    zIndex: 3,
    elevation: 3,
  },
  containerAlert: {
    width: '100%',
    padding: 8,
    borderRadius: 10,
    flexDirection: 'row',
    alignItems: 'flex-start',
    justifyContent: 'space-between',
  },
  wrapIcon: {
    marginTop: 2,
  },
  wrapText: {
    flex: 1,
    paddingHorizontal: 8,
  },
  textTitle: {
    ...FONT_HEADLINE_H6,
    color: COLORS.primaryWhite,
  },
  textDesc: {
    ...FONT_BODY5,
    lineHeight: 14,
    color: COLORS.primaryWhite,
    marginTop: 8,
  },
  wrapClose: {
    width: 20,
    borderRadius: 20 / 2,
  },
  containerClose: {
    width: '100%',
    height: 20,
    borderRadius: 20 / 2,
    padding: 0,
  },
});
