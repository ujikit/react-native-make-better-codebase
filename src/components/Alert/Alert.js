//package import
import React, { memo, useEffect, useState, useCallback } from 'react';
import { View, Text, Animated, ActivityIndicator } from 'react-native';
import PropTypes from 'prop-types';

//local import
import styles from './styles';
import { COLORS } from '../../configs';
import Button from '../Button';
import {
  CheckCircleIcon,
  WarningIcon,
  InfoCircleIcon,
  CloseCircleIcon,
  CrossIcon,
} from '../../assets/svgs';

const Alert = ({ close, desc, title, visible, types, duration, loading }) => {
  //state value
  const [bounceValue] = useState(new Animated.Value(-200));
  const [icon, setIcon] = useState(null);
  const [iconClose, setIconClose] = useState(null);
  const [color, setColor] = useState(COLORS.red50);

  //native effect
  useEffect(() => {
    if (visible) {
      _setVariable();
      _setIconClose();
      _show();
    }
  }, [_close, _setIconClose, _setVariable, _show, visible]);

  //place your function in here
  const _show = useCallback(() => {
    Animated.spring(bounceValue, {
      toValue: 0,
      velocity: 3,
      tension: 2,
      friction: 8,
      useNativeDriver: true,
    }).start();
    setTimeout(() => {
      _close();
    }, duration);
  }, [_close, bounceValue, duration]);

  const _close = useCallback(() => {
    close();
    Animated.spring(bounceValue, {
      toValue: -200,
      velocity: 3,
      tension: 2,
      friction: 8,
      useNativeDriver: true,
    }).start();
  }, [bounceValue, close]);

  const _setVariable = useCallback(() => {
    switch (types) {
      case 'success':
        setIcon(
          <CheckCircleIcon width="15" height="15" fill={COLORS.primaryWhite} />
        );
        setColor(COLORS.green50);
        break;
      case 'warning':
        setIcon(
          <WarningIcon width="15" height="15" fill={COLORS.primaryWhite} />
        );
        setColor(COLORS.orange50);
        break;
      case 'information':
        setIcon(
          <InfoCircleIcon width="15" height="15" fill={COLORS.primaryWhite} />
        );
        setColor(COLORS.blue50);
        break;
      case 'error':
        setIcon(
          <CloseCircleIcon width="15" height="15" fill={COLORS.primaryWhite} />
        );
        setColor(COLORS.red50);
        break;
    }
  }, [types]);

  const _setIconClose = useCallback(() => {
    setIconClose(
      <CrossIcon width="10" height="10" fill={COLORS.primaryWhite} />
    );
  }, []);

  return (
    <Animated.View
      style={[{ transform: [{ translateY: bounceValue }] }, styles.wrapAlert]}
    >
      <View style={[styles.containerAlert, { backgroundColor: color }]}>
        {loading ? (
          <ActivityIndicator color={COLORS.primaryOrange} />
        ) : (
          <View style={styles.wrapIcon}>{icon}</View>
        )}
        <View style={styles.wrapText}>
          {title ? (
            <Text numberOfLines={2} style={styles.textTitle}>
              {title}
            </Text>
          ) : null}
          {desc ? (
            <Text numberOfLines={2} style={styles.textDesc}>
              {desc}
            </Text>
          ) : null}
        </View>
        <Button
          types="nude"
          onPress={_close}
          styleWrap={styles.wrapClose}
          styleContainer={styles.containerClose}
        >
          {iconClose}
        </Button>
      </View>
    </Animated.View>
  );
};

Alert.propTypes = {
  types: PropTypes.oneOf(['success', 'warning', 'information', 'error']),
  visible: PropTypes.bool,
  title: PropTypes.string,
  desc: PropTypes.string,
  close: PropTypes.func,
  duration: PropTypes.number,
  loading: PropTypes.bool,
};

Alert.defaultProps = {
  types: 'error',
  visible: false,
  title: '',
  desc: '',
  close: () => {},
  duration: 3000,
  loading: false,
};

export default memo(Alert);
