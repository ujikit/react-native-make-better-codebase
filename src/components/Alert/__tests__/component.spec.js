import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Alert from '..';

describe('Some component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Alert />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
