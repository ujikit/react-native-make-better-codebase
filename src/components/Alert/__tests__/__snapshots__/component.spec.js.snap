// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Some component renders correctly 1`] = `
<View
  style={
    Object {
      "padding": 16,
      "position": "absolute",
      "top": null,
      "transform": Array [
        Object {
          "translateY": -200,
        },
      ],
      "width": "100%",
      "zIndex": 3,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "alignItems": "flex-start",
          "borderRadius": 10,
          "elevation": 3,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "padding": 8,
          "width": "100%",
        },
        Object {
          "backgroundColor": "#CF212A",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "marginTop": 2,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
          "paddingHorizontal": 8,
        }
      }
    />
    <View
      style={
        Array [
          Object {
            "borderRadius": 4,
            "overflow": "hidden",
            "width": null,
          },
          Object {
            "borderRadius": 10,
            "width": 20,
          },
        ]
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "transparent",
            "borderRadius": 10,
            "height": 20,
            "justifyContent": "center",
            "opacity": 1,
            "padding": 0,
            "width": "100%",
          }
        }
      />
    </View>
  </View>
</View>
`;
